src/blockchain/pop/vbk_block_tree.cpp

```diff
@@ -427,28 +427,79 @@ bool VbkBlockTree::loadTip(const Blob<24>& hash, ValidationState& state) {
   return true;
 }
 
-bool VbkBlockTree::areOnSameChain(const VbkBlock& blk1, const VbkBlock& blk2) {
-  auto* blk_index1 = this->getBlockIndex(blk1.getHash());
-  auto* blk_index2 = this->getBlockIndex(blk2.getHash());
+bool VbkBlockTree::areStronglyEquivalent(const VTB& vtb1, const VTB& vtb2) {
+  return (vtb1.transaction.bitcoinTransaction ==
+          vtb2.transaction.bitcoinTransaction) &&
+         (vtb1.transaction.blockOfProof == vtb2.transaction.blockOfProof);
+}
 
-  VBK_ASSERT_MSG(blk_index1, "unknown block %s", blk1.toPrettyString());
-  VBK_ASSERT_MSG(blk_index2, "unknown block %s", blk2.toPrettyString());
+bool VbkBlockTree::areWeaklyEquivalent(const VTB& vtb1, const VTB& vtb2) {
+  bool is_on_the_same_btc_chain = btc().areOnSameChain(
+      vtb1.transaction.blockOfProof, vtb2.transaction.blockOfProof);
 
-  if (blk_index1->getHeight() > blk_index2->getHeight()) {
-    return blk_index1->getAncestor(blk_index2->getHeight()) == blk_index2;
-  } else {
-    return blk_index2->getAncestor(blk_index1->getHeight()) == blk_index1;
-  }
+  bool is_on_the_same_vbk_chain = areOnSameChain(
+      vtb1.transaction.publishedBlock, vtb2.transaction.publishedBlock);
+
+  bool is_the_same_keystone_period =
+      abs(vtb1.transaction.publishedBlock.height -
+          vtb2.transaction.publishedBlock.height) <=
+      (int32_t)getParams().getKeystoneInterval();
+
+  return ((vtb1.transaction.publishedBlock ==
+           vtb2.transaction.publishedBlock) &&
+          is_on_the_same_btc_chain) ||
+         (is_on_the_same_vbk_chain && is_the_same_keystone_period &&
+          is_on_the_same_btc_chain);
 }
 
-bool VbkBlockTree::isStronglyEquivalent(const VTB& vtb1, const VTB& vtb2) {
-  bool are_on_the_same_chain =
-      areOnSameChain(vtb1.containingBlock, vtb2.containingBlock);
+int VbkBlockTree::weaklyCompare(const VTB& vtb1, const VTB& vtb2) {
+  VBK_ASSERT_MSG(areWeaklyEquivalent(vtb1, vtb2),
+                 "vtbs should be weakly equivalent");
 
-  return (vtb1.transaction.bitcoinTransaction ==
-          vtb2.transaction.bitcoinTransaction) &&
-         (vtb1.transaction.blockOfProof == vtb2.transaction.blockOfProof) &&
-         are_on_the_same_chain;
+  auto* blockOfProof1 =
+      btc().getBlockIndex(vtb1.transaction.blockOfProof.getHash());
+  auto* blockOfProof2 =
+      btc().getBlockIndex(vtb2.transaction.blockOfProof.getHash());
+
+  VBK_ASSERT_MSG(blockOfProof1,
+                 "unknown block %s",
+                 vtb1.transaction.blockOfProof.toPrettyString());
+  VBK_ASSERT_MSG(blockOfProof2,
+                 "unknown block %s",
+                 vtb2.transaction.blockOfProof.toPrettyString());
+
+  if (blockOfProof1->getHeight() > blockOfProof2->getHeight()) {
+    return 1;
```

Comment:
```suggestion
    return -1;
```
because 1st blockOfProof **is later**

Reply:
fixed